lib/rules/prefer-regex-literals.js

```diff
@@ -163,14 +257,83 @@ module.exports = {
                 };
 
                 for (const { node } of tracker.iterateGlobalReferences(traceMap)) {
-                    if (disallowRedundantWrapping && isUnnecessarilyWrappedRegexLiteral(node)) {
+                    scope = noStringScope(scope);
+
+                    if (disallowRedundantWrapping && isUnnecessarilyWrappedRegexLiteral(node, scope)) {
                         if (node.arguments.length === 2) {
-                            context.report({ node, messageId: "unexpectedRedundantRegExpWithFlags" });
+                            context.report({
+                                node,
+                                messageId: "unexpectedRedundantRegExpWithFlags",
+                                // eslint-disable-next-line no-loop-func -- scope value won't change
+                                fix(fixer) {
+                                    return fixer.replaceTextRange(node.range, node.arguments[0].raw + getStringValue(node.arguments[1], scope));
+                                }
+                            });
                         } else {
-                            context.report({ node, messageId: "unexpectedRedundantRegExp" });
+                            context.report({
+                                node,
+                                messageId: "unexpectedRedundantRegExp",
+                                fix(fixer) {
+                                    return fixer.replaceTextRange(node.range, node.arguments[0].raw);
+                                }
+                            });
+                        }
+                    } else if (
+                        (getStringValue(node.arguments[0], scope) !== null) &&
+                        (!node.arguments[1] || getStringValue(node.arguments[1], scope) !== null) &&
+                        (node.arguments.length === 1 || node.arguments.length === 2)
+                    ) {
+                        let regexContent = getStringValue(node.arguments[0], scope);
+                        let noFix = false;
+                        let flags;
+
+                        if (node.arguments[1]) {
+                            flags = getStringValue(node.arguments[1], scope);
                         }
-                    } else if (hasOnlyStaticStringArguments(node)) {
-                        context.report({ node, messageId: "unexpectedRegExp" });
+
+                        const RegExpValidatorInstance = new RegExpValidator({ ecmaVersion: getRegexppEcmaVersion(context.parserOptions.ecmaVersion) });
+
+                        try {
+                            RegExpValidatorInstance.validatePattern(regexContent);
```

Comment:
We should pass in the info on whether the `u` flag is present because it affects parsing and validation.